Multiple file change events

*Watchify* seems to be doing something *different* internally when writing its output stream than *other compilers*, for example

```js
var chokidar = require('chokidar');
var child = require('child_process');
var fs = require('fs');
var util = require('util');

var watcher = chokidar.watch('{output.js}', {
  alwaysStat: true
});

watcher.on('change', function (filename, stat) {
  console.log('change %s (size: %s)', filename, util.inspect(stat.size));
});

var ps = child.spawn('watchify', ['--outfile', 'output.js', 'index.js', '--verbose']);
ps.stdout.pipe(process.stdout);
ps.stderr.pipe(process.stderr);

setInterval(function() {
  console.log('touch');

  var input = fs.readFileSync('index.js', 'utf-8') + ' ';
  fs.writeFileSync('index.js', input, 'utf-8');
}, 1000);

setTimeout(function() {
  process.exit(0);
}, 10000);
```
```sh
$ node watchify.js 
touch
change output.js (size: 0)
124676 bytes written to output.js (0.04 seconds)
change output.js (size: 124676)
```

Where-as webpack in with the same use case results in a single change event
```js
var chokidar = require('chokidar');
var child = require('child_process');
var fs = require('fs');
var util = require('util');

var watcher = chokidar.watch('{output.js}', {
  alwaysStat: true
});

watcher.on('change', function (filename, stat) {
  console.log('change %s (size: %s)', filename, util.inspect(stat.size));
});

var ps = child.spawn('webpack', ['--watch', '--output-file', 'output.js', 'index.js']);
ps.stdout.pipe(process.stdout);
ps.stderr.pipe(process.stderr);

setInterval(function() {
  console.log('touch');

  var input = fs.readFileSync('index.js', 'utf-8') + ' ';
  fs.writeFileSync('index.js', input, 'utf-8');
}, 1000);

setTimeout(function() {
  process.exit(0);
}, 10000);
```
```sh
$ node webpack.js

touch
Hash: 0bd844fa25ecb086e1d9
Version: webpack 1.10.0
Time: 21ms
    Asset    Size  Chunks             Chunk Names
output.js  130 kB       0  [emitted]  main
   [0] ./index.js 132 bytes {0} [built]
    + 23 hidden modules
change output.js (size: 129664)
```

Treating size 0 as an edge case is somewhat trivial, but have also run into cases where change events are emitted 3-4 times when the file is in a fragmented state, babel, coffee and tsc don't seem to exhibit this behaviour either. Have not been able to reproduce this directly with fs.writeFile or fs.createWriteStream either.

Any idea why is watchify behaving so differently from the rest?

The proper solution I suppose would be to lock, but it's not in node core so I don't see that being widely adopted anytime soon, anyhow what I'm observing from *watchify* seems to be the following flow.

- dependency change
- wait 600
- write 0 bytes (externally observable)
- process output
- write pending bytes (externally observable)
- process output
- write pending bytes (externally observable)

The time I've observed so far between writes is anywhere between 25ms and 150ms.

For example, in [amok](https://github.com/caspervonb/amok)'s hot patching with watchify tests, which runs over babel, coffee, tsc, webpack and watchify, watchify is, so far the only one that has been causing issues and acting differently in terms of the writes (altho likely to have different results with Windows or Linux, as file watching is quite different per OS, I'll test that more in-depth later today)

The first issue, which we've been already discussing here is that there's an empty write, now this one is fairly easy to work around, you can't do anything with 0 bytes anyways and it should probably be treated as a file create anyhow instead of a file change. Working around it by just adding a one second  debounce delay, the next event will arrive and it will be like the first one did not occur at all.

The second manifestation is incomplete writes, it's all very inconsistent but sometimes there will only be 444 bytes available when reading, it's rare but it happens.

Those 444 bytes contain the prelude
```js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({
```

Atomic writes were so much easier to reason about..

--- a/bin/cmd.js
+++ b/bin/cmd.js
@@ -2,6 +2,7 @@
 
 var path = require('path');
 var outpipe = require('outpipe');
+var through = require('through2');
 
 var fromArgs = require('./args.js');
 var w = fromArgs(process.argv.slice(2));
@@ -33,24 +34,31 @@ bundle();
 
 function bundle () {
     var didError = false;
-    var outStream = outpipe(outfile);
-
+    var writer = through();
     var wb = w.bundle();
+    
+    w.pipeline.get('pack').once('readable', function() {
+        wb.pipe(writer);
+    });
+    
     wb.on('error', function (err) {
         console.error(String(err));
         didError = true;
-        outStream.end('console.error('+JSON.stringify(String(err))+');');
+        writer.end('console.error(' + JSON.stringify(String(err)) + ');');
     });
-    wb.pipe(outStream);
-
-    outStream.on('error', function (err) {
-        console.error(err);
-    });
-    outStream.on('exit', function () {
-        if (verbose && !didError) {
-            console.error(bytes + ' bytes written to ' + outfile
-                + ' (' + (time / 1000).toFixed(2) + ' seconds)'
-            );
-        }
+    
+    writer.once('readable', function() {
+        var outStream = outpipe(outfile);
+        outStream.on('error', function (err) {
+            console.error(err);
+        });
+        outStream.on('exit', function () {
+            if (verbose && !didError) {
+                console.error(bytes + ' bytes written to ' + outfile
+                    + ' (' + (time / 1000).toFixed(2) + ' seconds)'
+                );
+            }
+        });
+        writer.pipe(outStream);
     });
 }

